sdv/multi_table/base.py

```diff
@@ -0,0 +1,84 @@
+"""Base Multi Table Synthesizer class."""
+
+from collections import defaultdict
+from copy import deepcopy
+
+from sdv.single_table.copulas import GaussianCopulaSynthesizer
+
+
+class BaseMultiTableSynthesizer:
+    """Base class for multi table synthesizers.
+
+    The ``BaseMultiTableSynthesizer`` class defines the common API that all the
+    multi table synthesizers need to implement, as well as common functionality.
+
+    Args:
+        metadata (sdv.metadata.multi_table.MultiTableMetadata):
+            Multi table metadata representing the data tables that this synthesizer will be used
+            for.
+    """
+
+    _synthesizer = GaussianCopulaSynthesizer
+
+    def _initialize_models(self):
+        for table_name, table_metadata in self.metadata._tables.items():
+            synthesizer_parameters = self._table_parameters.get(table_name, {})
+            self._table_synthesizers[table_name] = self._synthesizer(
+                metadata=table_metadata,
+                **synthesizer_parameters
+            )
+
+    def __init__(self, metadata):
+        self.metadata = metadata
+        self.metadata.validate()
+        self._table_synthesizers = {}
+        self._table_parameters = defaultdict(dict)
+        self._initialize_models()
+
+    def get_table_parameters(self, table_name):
+        """Return the parameters that will be used to instantiate the table's synthesizer.
+
+        Args:
+            table_name (str):
+                Table name for which the parameters should be retrieved.
+
+        Returns:
+            parameters (dict):
+                A dictionary representing the parameters that will be used to instantiate the
+                table's synthesizer.
+        """
+        return self._table_parameters.get(table_name, {})
+
+    def get_parameters(self, table_name):
+        """Return the parameters used to instantiate the table's synthesizer.
+
+        Args:
+            table_name (str):
+                Table name for which the parameters should be retrieved.
+
+        Returns:
+            parameters (dict):
+                A dictionary representing the parameters used to instantiate the table's
+                synthesizer.
+        """
+        return self._table_synthesizers.get(table_name).get_parameters()
+
+    def update_table_parameters(self, table_name, table_parameters):
+        """Update the table's synthesizer instantiation parameters.
+
+        Args:
+            table_name (str):
+                Table name for which the parameters should be retrieved.
+            table_parameters (dict):
+                A dictionary with the parameters as keys and the values to be used to instantiate
+                the table's synthesizer.
+        """
+        self._table_parameters[table_name].update(deepcopy(table_parameters))
```

Comment:
can we move this line below the one under it in case the initialization crashes?